README.md
# ATtiny_Slow_PWM Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/ATtiny_Slow_PWM.svg?)](https://www.ardu-badge.com/ATtiny_Slow_PWM)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/ATtiny_Slow_PWM.svg)](https://github.com/khoih-prog/ATtiny_Slow_PWM/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/ATtiny_Slow_PWM/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/ATtiny_Slow_PWM.svg)](http://github.com/khoih-prog/ATtiny_Slow_PWM/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents

* [Important Note for Arduino IDE](#Important-Note-for-Arduino-IDE)
* [Why do we need this ATtiny_Slow_PWM library](#why-do-we-need-this-ATtiny_Slow_PWM-library)
  * [Features](#features)
  * [Why using ISR-based PWM is better](#why-using-isr-based-pwm-is-better)
  * [Currently supported Boards](#currently-supported-boards)
  * [Important Notes about ISR](#important-notes-about-isr)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [HOWTO Fix `Multiple Definitions` Linker Error](#howto-fix-multiple-definitions-linker-error)
* [More useful Information](#more-useful-information)
  * [1. Documents](#1-documents)
  * [2. Timer TCB0-TCB1](#2-timer-tcb0-tcb1)
* [Usage](#usage)
  * [1. Init Hardware Timer](#1-init-hardware-timer)
  * [2. Set PWM Frequency, dutycycle, attach irqCallbackStartFunc and irqCallbackStopFunc functions](#2-Set-PWM-Frequency-dutycycle-attach-irqCallbackStartFunc-and-irqCallbackStopFunc-functions)
* [Examples](#examples)
  * [ 1. ISR_8_PWMs_Array](examples/ISR_8_PWMs_Array)
  * [ 2. ISR_8_PWMs_Array_Complex](examples/ISR_8_PWMs_Array_Complex)
  * [ 3. ISR_8_PWMs_Array_Simple](examples/ISR_8_PWMs_Array_Simple)
  * [ 4. ISR_Changing_PWM](examples/ISR_Changing_PWM)
  * [ 5. ISR_Modify_PWM](examples/ISR_Modify_PWM)
  * [ 6. multiFileProject](examples/multiFileProject). **New**
* [Example ISR_8_PWMs_Array_Complex](#Example-ISR_8_PWMs_Array_Complex)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [1. ISR_8_PWMs_Array_Complex on AVR_ATtiny3217](#1-ISR_8_PWMs_Array_Complex-on-AVR_ATtiny3217)
  * [2. ISR_8_PWMs_Array on AVR_ATtiny3217](#2-isr_8_pwms_array-on-AVR_ATtiny3217)
  * [3. ISR_8_PWMs_Array_Simple on AVR_ATtiny3217](#3-ISR_8_PWMs_Array_Simple-on-AVR_ATtiny3217)
  * [4. ISR_Modify_PWM on AVR_ATtiny3217](#4-ISR_Modify_PWM-on-AVR_ATtiny3217)
  * [5. ISR_Changing_PWM on AVR_ATtiny3217](#5-ISR_Changing_PWM-on-AVR_ATtiny3217)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)


---
---

### Important Note for Arduino IDE

With some Arduino IDE versions, such as v1.8.19, upload directly via USB to some boards, such as `AVR_CuriosityNano3217` can't be done without `unknown-to-me` fix. We'll get the following error when uploading

```
avrdude: Version 6.3-20201216
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "/home/kh/.arduino15/packages/megaTinyCore/hardware/megaavr/2.5.11/avrdude.conf"
         User configuration file is "/home/kh/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping

         Using Port                    : usb
         Using Programmer              : curiosity_updi
avrdude: usbdev_open(): Found nEDBG CMSIS-DAP, serno: MCHP3333021800000998
avrdude: usbdev_open(): WARNING: failed to set configuration 1: Device or resource busy
avrdude: Found CMSIS-DAP compliant device, using EDBG protocol
avrdude: usbdev_send(): wrote -5 out of 912 bytes, err = Input/output error
avrdude: jtag3_edbg_prepare(): failed to send command to serial port

avrdude done.  Thank you.

the selected serial port 
 does not exist or your board is not connected
```

We can use drag-and-drop method to `drag-and-drop` the compiled **hex** file to `CURIOSITY` virtual drive. 

If `success`, The LED blinks **slowly** for 2 sec, or **rapidly** for 2 sec if `failure`


For example, to run [Change_Interval example](https://github.com/khoih-prog/ATtiny_Slow_PWM/tree/main/examples/Change_Interval), use Arduino IDE to compile, and get the `Change_Interval.ino.hex` file. For Ubuntu Linux, the file is stored in directory `/tmp/arduino_build_xxxxxx`


<p align="center">
    <img src="https://github.com/khoih-prog/ATtiny_Slow_PWM/raw/main/pics/Change_Interval.png">
</p>


After drag-and-drop the `Change_Interval.ino.hex` into `CURIOSITY` virtual drive, the code will run immediately if successfully loaded (LED blinks **slowly**)


<p align="center">
    <img src="https://github.com/khoih-prog/ATtiny_Slow_PWM/raw/main/pics/CURIOSITY_drive.png">
</p>


---
---

### Why do we need this [ATtiny_Slow_PWM library](https://github.com/khoih-prog/ATtiny_Slow_PWM)

### Features

This library enables you to use ISR-based PWM channels on Arduino AVR ATtiny-based boards (ATtiny3217, etc.) using [**megaTinyCore**](https://github.com/SpenceKonde/megaTinyCore), to create and output PWM any GPIO pin. Because this library doesn't use the powerful, high-speed, purely hardware-controlled PWM with **many limitations**, the maximum PWM frequency is currently limited at **1000Hz**, which is still suitable for many real-life applications, such as LED lighting, Servo, etc. Now you can change the PWM settings on-the-fly

---

This library enables you to use Interrupt from Hardware Timers on **AVR ATtiny-based boards** to create and output PWM to any pins. It now supports 64 ISR-based synchronized PWM channels, while consuming only 1 Hardware Timer. PWM interval can be very long (uint64_t microsecs / millisecs). The most important feature is they're `ISR-based PWM` channels. Therefore, their executions are not blocked by bad-behaving functions or tasks. This important feature is absolutely necessary for `mission-critical` tasks. These hardware PWM channels, using interrupt, still work even if other functions are blocking. Moreover, they are much more precise (certainly depending on clock frequency accuracy) than other software PWM using millis() or micros(). That's necessary if you need to measure some data requiring better accuracy.

As **Hardware Timers are rare, and very precious assets** of any board, this library now enables you to use up to **64 ISR-based synchronized PWM channels, while consuming only 1 Hardware Timer**. Timers' interval is very long (**ulong millisecs**).

Now with these new **64 ISR-based PWM-channels**, the maximum interval is **practically unlimited** (limited only by `uint64_t` milliseconds) while **the accuracy is nearly perfect** compared to software PWM channels. 

The most important feature is they're `ISR-based PWM` channels. Therefore, their executions are **not blocked by bad-behaving functions / tasks**, such as connecting to WiFi, Internet or Blynk services. This important feature is absolutely necessary for mission-critical tasks. You can also have many `(up to 64)` PWM channels to use.

This non-being-blocked important feature is absolutely necessary for mission-critical tasks.

The [**ISR_8_PWMs_Array_Complex**](examples/ISR_8_PWMs_Array_Complex) example will demonstrate the nearly perfect accuracy, compared to software PWM, by printing the actual period / duty-cycle in `microsecs` of each of PWM-channels.

You'll see `software-based` SimpleTimer is blocked while system is connecting to WiFi / Internet / Blynk, as well as by blocking task in loop(), using delay() function as an example. The elapsed time then is very unaccurate

---

#### Why using ISR-based PWM is better

Imagine you have a system with a **mission-critical** function, measuring water level and control the sump pump or doing something much more important. You normally use a software timer to poll, or even place the function in loop(). But what if another function is **blocking** the loop() or setup().

So your function **might not be executed, and the result would be disastrous.**

You'd prefer to have your function called, no matter what happening with other functions (busy loop, bug, etc.).

The correct choice is to use a Hardware Timer with **Interrupt** to call your function.

These hardware PWM channels, using interrupt, still work even if other functions are blocking. Moreover, they are much more **precise** (certainly depending on clock frequency accuracy) than other software PWM channels using millis() or micros(). That's necessary if you need to measure some data requiring better accuracy.

Functions using normal software PWM channels, relying on loop() and calling millis(), won't work if the loop() or setup() is blocked by certain operation. For example, certain function is blocking while it's connecting to WiFi or some services.

The catch is **your function is now part of an ISR (Interrupt Service Routine), and must be lean / mean, and follow certain rules.** More to read on:

[**HOWTO Attach Interrupt**](https://www.arduino.cc/reference/en/language/functions/external-interrupts/attachinterrupt/)

---

### Currently supported Boards

- **tinyAVR boards using megaTinyCore**

##### Curiosity Nano ATtiny3217

<p align="center">
    <img src="https://github.com/khoih-prog/ATtiny_TimerInterrupt/raw/main/pics/Curiosity_ATtiny3217.png">
</p>


---


#### Important Notes about ISR

1. Inside the attached function, **delay() won’t work and the value returned by millis() will not increment.** Serial data received while in the function may be lost. You should declare as **volatile any variables that you modify within the attached function.**

2. Typically global variables are used to pass data between an ISR and the main program. To make sure variables shared between an ISR and the main program are updated correctly, declare them as `volatile`.

---
---

## Prerequisites

1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
2. [`SpenceKonde megaTinyCore core 2.6.4+`](https://github.com/SpenceKonde/megaTinyCore) for Arduino ATtiny boards.  [![GitHub release](https://img.shields.io/github/release/SpenceKonde/megaTinyCore.svg)](https://github.com/SpenceKonde/megaTinyCore/releases/latest). Follow [**megaTinyCore Installation**](https://github.com/SpenceKonde/megaTinyCore/blob/master/Installation.md).
3. To use with certain example
   - [`SimpleTimer library`](https://github.com/jfturcot/SimpleTimer) for [ISR_8_PWMs_Array_Simple](examples/ISR_8_PWMs_Array_Simple) and [ISR_8_PWMs_Array_Complex](examples/ISR_8_PWMs_Array_Complex) examples.
   
   
---
---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for [**ATtiny_Slow_PWM**](https://github.com/khoih-prog/ATtiny_Slow_PWM), then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/ATtiny_Slow_PWM.svg?)](https://www.ardu-badge.com/ATtiny_Slow_PWM) for more detailed instructions.

### Manual Install

Another way to install is to:

1. Navigate to [**ATtiny_Slow_PWM**](https://github.com/khoih-prog/ATtiny_Slow_PWM) page.
2. Download the latest release `ATtiny_Slow_PWM-main.zip`.
3. Extract the zip file to `ATtiny_Slow_PWM-main` directory 
4. Copy whole `ATtiny_Slow_PWM-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**ATtiny_Slow_PWM** library](https://registry.platformio.org/libraries/khoih-prog/ATtiny_Slow_PWM) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/ATtiny_Slow_PWM/installation). Search for **ATtiny_Slow_PWM** in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)


---
---

### HOWTO Fix `Multiple Definitions` Linker Error

The current library implementation, using `xyz-Impl.h` instead of standard `xyz.cpp`, possibly creates certain `Multiple Definitions` Linker error in certain use cases.

You can include this `.hpp` file

```cpp
// Can be included as many times as necessary, without `Multiple Definitions` Linker Error
#include "ATtiny_Slow_PWM.hpp"          //https://github.com/khoih-prog/ATtiny_Slow_PWM
```

in many files. But be sure to use the following `.h` file **in just 1 `.h`, `.cpp` or `.ino` file**, which must **not be included in any other file**, to avoid `Multiple Definitions` Linker Error

```cpp
// To be included only in main(), .ino with setup() to avoid `Multiple Definitions` Linker Error
#include "ATtiny_Slow_PWM.h"            //https://github.com/khoih-prog/ATtiny_Slow_PWM
```

Check the new [**multiFileProject** example](examples/multiFileProject) for a `HOWTO` demo.

Have a look at the discussion in [Different behaviour using the src_cpp or src_h lib #80](https://github.com/khoih-prog/ESPAsync_WiFiManager/discussions/80)

---
---

## More useful Information

### 1. Documents

1. [Arduino 101: Timers and Interrupts](https://www.robotshop.com/community/forum/t/arduino-101-timers-and-interrupts/13072)
2. [Getting Started with Timer/Counter Type B (TCB)](https://ww1.microchip.com/downloads/aemDocuments/documents/MCU08/ApplicationNotes/ApplicationNotes/TB3214-Getting-Started-with-TCB-DS90003214.pdf)
3. [megaTinyCore README.md](https://github.com/SpenceKonde/megaTinyCore/blob/master/README.md)
4. [ATtiny3217 Curiosity Nano Hardware User Guide](https://ww1.microchip.com/downloads/en/DeviceDoc/40002193A.pdf)
5. [AVR128DA48-Curiosity-Nano-Hardware-User Guide](https://ww1.microchip.com/downloads/en/DeviceDoc/AVR128DA48-Curiosity-Nano-UG-DS50002971B.pdf)
6. [AVR128DB48-Curiosity-Nano-Hardware-User Guide](https://ww1.microchip.com/downloads/en/DeviceDoc/AVR128DB48-Curiosity-Nano-HW-UserG-DS50003037A.pdf)


### 2. Timer TCB0-TCB1

TCB0-TCB1 are 16-bit timers

The ATtiny boards, such as `ATtiny3217`, `ATtiny1617`, will have only maximum 2 TCB timers, (TCB0-TCB1).

The ATtiny boards, such as `ATtiny817`, will have only maximum 1 TCB timer, (TCB0).

The number of TCB timers will be automatically configured by the library.

The following is the partial list of number of TCBs for each ATtiny board/chip

#### TCB0-TCB1, TCA, TCD

**ATtiny3217, ATtiny1617, ATtiny3216, ATtiny1616, ATtiny1614**

#### TCB0, TCA, TCD

**ATtinyx12, ATtinyx14, ATtinyx16, ATtinyx17, such as ATtiny817, ATtiny417, ATtiny816, etc.**

#### TCB0, TCA, no TCD

**ATtinyx02, ATtinyx04, ATtinyx06, ATtinyx07, such as ATtiny1607, ATtiny807, ATtiny1606, etc.**

---
---

## Usage


Before using any Timer, you have to make sure the Timer has not been used by any other purpose.


#### 1. Init Hardware Timer

```cpp
// Select USING_FULL_CLOCK      == true for  20/16MHz to Timer TCBx => shorter timer, but better accuracy
// Select USING_HALF_CLOCK      == true for  10/ 8MHz to Timer TCBx => shorter timer, but better accuracy
// Select USING_250KHZ          == true for 250KHz to Timer TCBx => longer timer,  but worse  accuracy
// Not select for default 250KHz to Timer TCBx => longer timer,  but worse accuracy
#define USING_FULL_CLOCK      true
#define USING_HALF_CLOCK      false
#define USING_250KHZ          false         // Not supported now

// Try to use RTC, TCA0 or TCD0 for millis()
#define USE_TIMER_0           true          // Check if used by millis(), Servo or tone()
#define USE_TIMER_1           false         // Check if used by millis(), Servo or tone()

#if USE_TIMER_0
  #define CurrentTimer   ITimer0
#elif USE_TIMER_1
  #define CurrentTimer   ITimer1
#else
  #error You must select one Timer  
#endif

// To be included only in main(), .ino with setup() to avoid `Multiple Definitions` Linker Error
#include "ATtiny_Slow_PWM.h"


changelog.md
## ATtiny_Slow_PWM Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/ATtiny_Slow_PWM.svg?)](https://www.ardu-badge.com/ATtiny_Slow_PWM)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/ATtiny_Slow_PWM.svg)](https://github.com/khoih-prog/ATtiny_Slow_PWM/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/ATtiny_Slow_PWM/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/ATtiny_Slow_PWM.svg)](http://github.com/khoih-prog/ATtiny_Slow_PWM/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Initial Release v1.0.0](#initial-release-v100)

---
---

## Changelog


### Initial Release v1.0.0

1. Initial release to support Arduino **AVR ATtiny-based boards (ATtiny3217, etc.) using megaTinyCore**





src/ATtiny_Slow_PWM_Impl.h
/****************************************************************************************************************************
  ATtiny_Slow_PWM_Impl.h
  For Arduino AVR ATtiny-based boards (ATtiny3217, etc.) using megaTinyCore
  Written by Khoi Hoang

  Built by Khoi Hoang https://github.com/khoih-prog/ATtiny_Slow_PWM
  Licensed under MIT license

  Now even you use all these new 16 ISR-based timers,with their maximum interval practically unlimited (limited only by
  unsigned long miliseconds), you just consume only one AVRDx-based timer and avoid conflicting with other cores' tasks.
  The accuracy is nearly perfect compared to software timers. The most important feature is they're ISR-based timers
  Therefore, their executions are not blocked by bad-behaving functions / tasks.
  This important feature is absolutely necessary for mission-critical tasks.

  Version: 1.0.0

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0   K.Hoang      27/08/2022 Initial coding to support AVR ATtiny (ATtiny3217, etc.) using megaTinyCore
****************************************************************************************************************************/

#pragma once

#ifndef AT_TINY_SLOW_PWM_IMPL_H
#define AT_TINY_SLOW_PWM_IMPL_H

#ifndef _PWM_LOGLEVEL_
  #define _PWM_LOGLEVEL_      0
#endif

#include <string.h>

/*****************************************************************************************

  // From ~/.arduino15/packages/arduino/7.3.0-atmel3.6.1-arduino5/avr/include/avr/iom4809.h

  //#define TCB0                  (*(TCB_t *) 0x0A80) // 16-bit Timer Type B
  //#define TCB1                  (*(TCB_t *) 0x0A90) // 16-bit Timer Type B
  //#define TCB2                  (*(TCB_t *) 0x0AA0) // 16-bit Timer Type B
  //#define TCB3                  (*(TCB_t *) 0x0AB0) // 16-bit Timer Type B

  //
  typedef enum TCB_CLKSEL_enum
  {
    TCB_CLKSEL_CLKDIV1_gc = (0x00<<1),  // CLK_PER (No Prescaling)
    TCB_CLKSEL_CLKDIV2_gc = (0x01<<1),  // CLK_PER/2 (From Prescaler)
    TCB_CLKSEL_CLKTCA_gc = (0x02<<1),   // Use Clock from TCA
  } TCB_CLKSEL_t;

  //
  typedef enum TCB_CNTMODE_enum
  {
    TCB_CNTMODE_INT_gc = (0x00<<0),       // Periodic Interrupt
    TCB_CNTMODE_TIMEOUT_gc = (0x01<<0),   // Periodic Timeout
    TCB_CNTMODE_CAPT_gc = (0x02<<0),      // Input Capture Event
    TCB_CNTMODE_FRQ_gc = (0x03<<0),       // Input Capture Frequency measurement
    TCB_CNTMODE_PW_gc = (0x04<<0),        // Input Capture Pulse-Width measurement
    TCB_CNTMODE_FRQPW_gc = (0x05<<0),     // Input Capture Frequency and Pulse-Width measurement
    TCB_CNTMODE_SINGLE_gc = (0x06<<0),    // Single Shot
    TCB_CNTMODE_PWM8_gc = (0x07<<0),      // 8-bit PWM
  } TCB_CNTMODE_t;

*****************************************************************************************/

/*****************************************************************************************

  // From https://github.com/SpenceKonde/DxCore/blob/master/megaavr/cores/dxcore/Arduino.h#L455-L462

  #define TIMERA0         (0x10) // A "simple" type A timer mapping doesn't get constants for the WO channels.
  #define TIMERA1         (0x08) // Formerly 0x11 - giving it a dedicated bit makes the takeover tracking easy and efficient instead of being a morass of tests and bitmath.
  #define TIMERB0         (0x20) // TCB0
  #define TIMERB1         (0x21) // TCB1
  #define TIMERB2         (0x22) // TCB2
  #define TIMERB3         (0x23) // TCB3
  #define TIMERB4         (0x24) // TCB4
  #define TIMERD0         (0x70) // If any of these bits match it's potentially on TCD0

*****************************************************************************************/

#if defined(TCB1)
#if (_TIMERINTERRUPT_LOGLEVEL_ > 2)
  #warning TCB0-TCB1 Timers available
#endif

TCB_t* TimerTCB[ NUM_HW_TIMERS ] = { &TCB0, &TCB1 };

#else
#if (_TIMERINTERRUPT_LOGLEVEL_ > 2)
  #warning Only TCB0 Timer available
#endif

TCB_t* TimerTCB[ NUM_HW_TIMERS ] = { &TCB0 };

#endif

//////////////////////////////////////////////

// Assuming Clock is 16MHz
#if USING_FULL_CLOCK
  // Use no prescaler (prescaler 1) => 20/16MHz, etc.
  #if (_PWM_LOGLEVEL_ > 2)
    #warning Using no prescaler => FULL_CLOCK (20MHz, 16MHz, etc)
  #endif

  #define TCB_CLKSEL_VALUE      TCB_CLKSEL_DIV1_gc
  #define CLOCK_PRESCALER       1

#elif USING_HALF_CLOCK
  // Use prescaler 2 => 12/8MHz, etc.
  #if (_PWM_LOGLEVEL_ > 2)
    #warning Using prescaler 2 => HALF_CLOCK (10MHz, 8MHz, etc.)
  #endif

  #define TCB_CLKSEL_VALUE      TCB_CLKSEL_DIV2_gc
  #define CLOCK_PRESCALER       2

  // No TCB_CLKSEL_CLKTCA_gc in megaTinyCore ???
  // Don't use 250KHz now
#elif USING_250KHZ
  // Optional, but for clarity
  // Use Timer A as clock (prescaler 64) => 250KHz
  #define TCB_CLKSEL_VALUE      TCB_CLKSEL_CLKTCA_gc
  #define CLOCK_PRESCALER       64

  #error Not OK now. Do not use

#else
  // Use prescaler 2 => 10/8MHz
  #if (_PWM_LOGLEVEL_ > 2)
    #warning Using prescaler 2 => HALF_CLOCK (10MHz, 8MHz, etc.)
  #endif

  #define TCB_CLKSEL_VALUE      TCB_CLKSEL_DIV2_gc
  #define CLOCK_PRESCALER       2

#endif

#define CLK_TCB_FREQ          ( F_CPU / CLOCK_PRESCALER )

//////////////////////////////////////////////

void TimerInterrupt::init(const int8_t& timer)
{
  // Set timer specific stuff
  // All timers in CTC mode
  // 8 bit timers will require changing prescalar values,
  // whereas 16 bit timers are set to either ck/1 or ck/64 prescalar

  noInterrupts();

  // 16 bit timer
  TimerTCB[timer]->CTRLB    = TCB_CNTMODE_INT_gc;                         // Use timer compare mode
  TimerTCB[timer]->CCMP     = MAX_COUNT_16BIT;                            // Value to compare with.
  TimerTCB[timer]->INTCTRL  &= ~TCB_CAPT_bm;                              // Disable the interrupt
  TimerTCB[timer]->CTRLA    = TCB_CLKSEL_VALUE | TCB_ENABLE_bm;           // Use Timer A as clock, enable timer

  _timer = timer;

  interrupts();
}

//////////////////////////////////////////////

void TimerInterrupt::set_CCMP()
{
  // Run with noInterrupt()
  // Set the CCMP for the given timer,
  // set the toggle count,
  // then turn on the interrupts
  uint32_t _CCMPValueToUse;

  _CCMPValueToUse = min(MAX_COUNT_16BIT, _CCMPValueRemaining);
  _CCMPValueRemaining -= _CCMPValueToUse;

  TimerTCB[_timer]->CCMP    = _CCMPValueToUse;    // Value to compare with.

  TimerTCB[_timer]->INTCTRL = TCB_CAPT_bm; // Enable the interrupt

  // Flag _CCMPValue == 0 => end of long timer
  if (_CCMPValueRemaining == 0)
    _timerDone = true;

}

//////////////////////////////////////////////

// frequency (in hertz) and duration (in milliseconds).
// Return true if frequency is OK with selected timer (CCMPValue is in range)
bool TimerInterrupt::setFrequency(const float& frequency, timer_callback_p callback, const uint32_t& params,
                                  const unsigned long& duration)
{
  //frequencyLimit must > 1
  float frequencyLimit = frequency * 17179.840;

  // Limit frequency to larger than (0.00372529 / 64) Hz or interval 17179.840s / 17179840 ms to avoid uint32_t overflow
  if ((_timer < 0) || (callback == NULL) || ((frequencyLimit) < 1) )
  {
    return false;
  }
  else
  {
    // Calculate the toggle count. Duration must be at least longer then one cycle
    if (duration > 0)
    {
      _toggle_count = frequency * duration / 1000;

      if (_toggle_count < 1)
      {
        return false;
      }
    }
    else
    {
      _toggle_count = -1;
    }

    //Timer0-4 are 16 bit timers, meaning it can store a maximum counter value of 65535.

    noInterrupts();

    _frequency = frequency;
    _callback  = (void*) callback;
    _params    = reinterpret_cast<void*>(params);

    _timerDone = false;

    _CCMPValue = _CCMPValueRemaining = (uint32_t) (CLK_TCB_FREQ / frequency);

    // Set the CCMP for the given timer,
    // set the toggle count,
    // then turn on the interrupts
    set_CCMP();

    interrupts();

    return true;
  }
}

//////////////////////////////////////////////

void TimerInterrupt::detachInterrupt()
{
  noInterrupts();

  // Clear interrupt flag
  TimerTCB[_timer]->INTFLAGS = TCB_CAPT_bm;
  TimerTCB[_timer]->INTCTRL  &= ~TCB_CAPT_bm;    // Disable the interrupt
  TimerTCB[_timer]->CTRLA    &= ~TCB_ENABLE_bm;  // Disable timer

  interrupts();
}

//////////////////////////////////////////////

// Duration (in milliseconds). Duration = 0 or not specified => run indefinitely
void TimerInterrupt::reattachInterrupt(const unsigned long& duration)
{
  noInterrupts();

  // Calculate the toggle count
  if (duration > 0)
  {
    _toggle_count = _frequency * duration / 1000;
  }
  else
  {
    _toggle_count = -1;
  }

  // Set interrupt flag
  TimerTCB[_timer]->INTCTRL  |= TCB_CAPT_bm;    // Enable the interrupt
  TimerTCB[_timer]->CTRLA    |= TCB_ENABLE_bm;  // Enable timer

  interrupts();
}

//////////////////////////////////////////////

// Just stop clock source, still keep the count
// To fix this.
void TimerInterrupt::pauseTimer()
{
  detachInterrupt();
}

//////////////////////////////////////////////

// Just reconnect clock source, continue from the current count
void TimerInterrupt::resumeTimer()
{
  reattachInterrupt();
}

////////////////////////////////////////////////////////

// To be sure not used Timers are disabled

// TCB0
#if !defined(USE_TIMER_0)
  #define USE_TIMER_0     false
#endif

// TCB1
#if !defined(USE_TIMER_1)
  #define USE_TIMER_1     false
#endif

////////////////////////////////////////////////////////////////////////////////

#if USE_TIMER_0
#ifndef TIMER0_INSTANTIATED
// To force pre-instantiate only once
#define TIMER0_INSTANTIATED
TimerInterrupt ITimer0(HW_TIMER_0);

ISR(TCB0_INT_vect)
{
  long countLocal = ITimer0.getCount();

  if (ITimer0.getTimer() == 0)
  {
    if (countLocal != 0)
    {
      if (ITimer0.checkTimerDone())
      {
        ITimer0.callback();

        if (ITimer0.get_CCMPValue() > MAX_COUNT_16BIT)
        {
          // To reload _CCMPValueRemaining as well as _CCMP register to MAX_COUNT_16BIT
          ITimer0.reload_CCMPValue();
        }

        if (countLocal > 0)
          ITimer0.setCount(countLocal - 1);
      }
      else
      {
        //Deduct _CCMPValue by min(MAX_COUNT_16BIT, _CCMPValue)
        // If _CCMPValue == 0, flag _timerDone for next cycle
        // If last one (_CCMPValueRemaining < MAX_COUNT_16BIT) => load _CCMP register _CCMPValueRemaining
        ITimer0.adjust_CCMPValue();
      }
    }
    else
    {
      ITimer0.detachInterrupt();
    }
  }

  // Clear interrupt flag
  TCB0.INTFLAGS = TCB_CAPT_bm;
}
#endif  //#ifndef TIMER0_INSTANTIATED
#endif    //#if USE_TIMER_0

//////////////////////////////////////////////

#if USE_TIMER_1
#ifndef TIMER1_INSTANTIATED
// To force pre-instantiate only once
#define TIMER1_INSTANTIATED
TimerInterrupt ITimer1(HW_TIMER_1);

// Timer0 is used for micros(), millis(), delay(), etc and can't be used
// Pre-instantiate

ISR(TCB1_INT_vect)
{
  long countLocal = ITimer1.getCount();

  if (ITimer1.getTimer() == 1)
  {
    if (countLocal != 0)
    {
      if (ITimer1.checkTimerDone())
      {
        ITimer1.callback();

        if (ITimer1.get_CCMPValue() > MAX_COUNT_16BIT)
        {
          // To reload _CCMPValueRemaining as well as _CCMP register to MAX_COUNT_16BIT if _CCMPValueRemaining > MAX_COUNT_16BIT
          ITimer1.reload_CCMPValue();
        }

        if (countLocal > 0)
          ITimer1.setCount(countLocal - 1);
      }
      else
      {
        //Deduct _CCMPValue by min(MAX_COUNT_16BIT, _CCMPValue)
        // If _CCMPValue == 0, flag _timerDone for next cycle
        // If last one (_CCMPValueRemaining < MAX_COUNT_16BIT) => load _CCMP register _CCMPValueRemaining
        ITimer1.adjust_CCMPValue();
      }
    }
    else
    {
      ITimer1.detachInterrupt();
    }
  }

  // Clear interrupt flag
  TCB1.INTFLAGS = TCB_CAPT_bm;
}

#endif  //#ifndef TIMER1_INSTANTIATED
#endif    //#if USE_TIMER_1

//////////////////////////////////////////////

#endif // AT_TINY_SLOW_PWM_IMPL_H
